src/remote_tables/pushdown_visitor.rs

```
@@ -0,0 +1,204 @@
+                });
+                let mut left_sql = self
+                    .sql_exprs
+                    .pop()
+                    .unwrap_or_else(|| panic!("Missing left sub-expression of {}", expr));
+
+                // Similar as in Display impl for BinaryExpr: since the Expr has an implicit operator
+                // precedence we need to convert it to an explicit one using extra parenthesis if the
+                // left/right expression is also a BinaryExpr of lower operator precedence.
+                if let Expr::BinaryExpr(right_be @ BinaryExpr { .. }) = &*be.right {
+                    let p = right_be.precedence();
+                    if p == 0 || p < be.precedence() {
+                        right_sql = format!("({})", right_sql)
+                    }
+                }
+                if let Expr::BinaryExpr(left_be @ BinaryExpr { .. }) = &*be.left {
+                    let p = left_be.precedence();
+                    if p == 0 || p < be.precedence() {
+                        left_sql = format!("({})", left_sql)
+                    }
+                }
+
+                let op = self.source.op_to_sql(be.op).ok_or_else(|| {
+                    DataFusionError::Execution(format!(
+                        "Couldn't convert operator {:?} to a compatible one for the remote system",
+                        be.op,
+                    ))
+                })?;
+
+                self.sql_exprs
+                    .push(format!("{} {} {}", left_sql, op, right_sql))
+            }
+            _ => {}
+        };
+        Ok(self)
+    }
+}
+
+// Walk the filter expression AST for a particular remote source type and see if the expression is
+// ship-able, at the same time converting elements (e.g. operators) to the native representation if
+// needed.
+pub fn filter_expr_to_sql<T: FilterPushdownVisitor>(
+    filter: &Expr,
+    source_pushdown: T,
+) -> Result<String> {
```

Currently it looks like an `Err(...Execution)` is going to be silently swallowed by `filter_map`, so we can't distinguish between pushdowns being unsupported for a certain operator and genuine errors (if those can be raised by this function). Should this be an Option instead, if this function can't error out?

---

I prefer keeping it `Result<String>` so as to propagate.expose that and any visitor errors. I've added a change where this is used for logging.